[WIP] Consolidated 4.10 RAN docs (#43890)
openshift/openshift-docs: modules/cnf-topology-aware-lifecycle-operator-preparing-for-updates.adoc

// Module included in the following assemblies:
// Epic CNF-2600 (CNF-2133) (4.10), Story TELCODOCS-285
// * scalability_and_performance/ztp-deploying-disconnected.adoc

:_content-type: PROCEDURE
[id="talo-platform-prepare-end-to-end_{context}"]
= End-to-end procedures for updating clusters in a disconnected environment

If you have deployed spoke clusters with distributed unit (DU) profiles using the GitOps ZTP with the {cgu-operator-first} pipeline described in "Deploying distributed units at scale in a disconnected environment", this procedure describes how to upgrade your spoke clusters and Operators.

[id="talo-platform-prepare-for-update_{context}"]
== Preparing for the updates

> If both the hub and the spoke clusters are running {product-title} 4.9, you must update ZTP from version 4.9 to 4.10. If {product-title} 4.10 is used, you can set up the environment.

Reviewer: Sentence 
"If OpenShift Container Platform 4.10 is used, you can set up the environment." 
is not clear to me.
 OCP 4.10 where? Spoke? Hub?
Also "set up the environment" is confusing. How about "proceed to next section"?

----
$ cat >checksum-${OCP_RELEASE_NUMBER}.yaml <<EOF
${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATURE_BASE64}
EOF
----

. Prepare the update graph. You have two options to prepare the update graph:

.. Use the OpenShift Update Service.
+
For more information about how to set up the graph on the hub cluster, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#deploy-the-operator-for-cincinnati[Deploy the operator for OpenShift Update Service] and link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#build-the-graph-data-init-container[Build the graph data init container].

.. Make a local copy of the upstream graph. Host the update graph on an `http` or `https` server in the disconnected environment that has access to the spoke cluster. To download the update graph, use the following command:
+
[source,terminal]
----
$ curl -s https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.10 -o ~/upgrade-graph_stable-4.10
----

* For Operator updates, you must perform the following tasks:
+

> . Mirror the Operator catalogs. Ensure that the desired operator images are mirrored by following the procedure in the "Mirroring Operator catalogs for use with disconnected clusters" section.

Reviewer 2: For a single-step procedure, use a bullet rather than number.
```suggestion
* Mirror the Operator catalogs. Ensure that the desired operator images are mirrored by following the procedure in the "Mirroring Operator catalogs for use with disconnected clusters" section.
```

TALO can perform both platform and Operator updates:

* For platform updates, you have to do the following:
+
. Mirror the desired {product-title} image repository.
+ 
For more information, see xref:../scalability_and_performance/ztp-deploying-disconnected.adoc#installation-mirror-repository_ztp-deploying-disconnected[Mirroring the OpenShift Container Platform image repository]
+
. Prepare the update graph. You can do this the following ways:
.. Use the OpenShift Update Service.
+
For more information, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/latest/html/clusters/managing-your-clusters#upgrading-disconnected-clusters[Upgrading disconnected clusters].
.. Make a local copy of the upstream graph and save the graph in the disconnected environment.

* For Operator updates, you have to do the following:
+

> . Mirror the desired index image, bundle images, and all Operator images referenced in the bundle images.

Author: @yliu127 @Missxiaoguo The _Preparing the disconnected environment_ section and related links in TALO have been updated:
https://deploy-preview-43890--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/ztp-deploying-disconnected.html#ztp-acm-preparing-to-install-disconnected-acm_ztp-deploying-disconnected

// Module included in the following assemblies:
// Epic CNF-2600 (CNF-2133) (4.10), Story TELCODOCS-285
// * scalability_and_performance/ztp-deploying-disconnected.adoc

:_content-type: PROCEDURE
[id="talo-platform-prepare-end-to-end_{context}"]

> = End-to-end procedures for updating clusters in a disconnected environment

Reviewer 3: Add a short description something like the following

If you have deployed your spoke clusters with distributed unit(DU) profiles using the GitOps ZTP with Topology Aware Lifecycle Operator pipeline described in "Deploying distributed units at scale in a disconnected environment", this procedure tells you how to upgrade your spoke clusters and Operators.

// Module included in the following assemblies:
// Epic CNF-2600 (CNF-2133) (4.10), Story TELCODOCS-285
// * scalability_and_performance/ztp-deploying-disconnected.adoc

:_content-type: PROCEDURE
[id="talo-platform-prepare-end-to-end_{context}"]
= End-to-end procedures for updating clusters in a disconnected environment

If you have deployed spoke clusters with distributed unit (DU) profiles using the GitOps ZTP with Topology Aware Lifecycle Operator pipeline described in "Deploying distributed units at scale in a disconnected environment", this procedure tells you how to upgrade your spoke clusters and Operators.

[id="talo-platform-prepare-for-update_{context}"]
== Preparing for the updates

If both the hub and the spoke clusters are running {product-title} 4.9, you need to update ZTP from version 4.9 to 4.10. If {product-title} 4.10 is used, you can set up the environment.

[id="talo-platform-prepare-for-update-env-setup_{context}"]

> == Setting up the environment

Reviewer 3: Hey Alexandra, this section is not clearly described. Please check out the Google doc I updated for this whole section.

If you deployed ZTP in a disconnected environment as instructed in the "Preparing the disconnected environment" section in "Deploying distributed units at scale in a disconnected environment", you need to do the following steps before TALO can successfully perform update operations:

TALO can perform both platform and Operator updates:

* For platform updates, you have to do the following:
+
. Mirror the desired {product-title} image repository. For more information about how to mirror an image repository, see the "Mirroring the OpenShift Container Platform image repository" section.
+
[IMPORTANT]
====

> The `imageContentSources` section, which can be retrieved with the `oc adm release mirror` command, needs to be saved for platform updates.
> For example, this is the output of the `imageContentSources` which contains the information of your mirrors:

Reviewer 2: We shouldn't be using commands in admonition blocks as they have the risk of getting missed in the sequence. Can this be restructured as a step?
```suggestion
The `imageContentSources` section, which can be retrieved with the `oc adm release mirror` command, must be saved for platform updates.
For example, this is the output of the `imageContentSources` which contains the information of your mirrors:
```

[source,terminal]
----
imageContentSources:
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-release
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
----
====

. Prepare the update graph. You have two options to prepare the update graph:

.. Use the OpenShift Update Service.
+

> For more information about how to set up the graph on the hub cluster, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#build-the-graph-data-init-container[Build the graph data init container].

Reviewer 2: You've been initializing "Hub cluster" elsewhere, should it be here?

Reviewer 4: @bobfuru @amolnar-rh @aireilly   "hub cluster" is lowercase.

Author: @bobfuru @StephenJamesSmith  I'm going through the files and correcting it!

> .. Make a local copy of the upstream graph and host the update graph on a http or https server in the disconnected environment to which the spoke cluster has access. To download the update graph, use the following command:

Reviewer 2: This first sentence is pretty confusing. It mentions an upstream graph and update graph, and the "to which" wording is ambiguous.
```suggestion
.. Make a local copy of the upstream graph. Host the update graph on an `http` or `https` server in the disconnected environment that has access to the spoke cluster. To download the update graph, use the following command:
```

+
[source,terminal]
----
$ curl -s https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.10 -o ~/upgrade-graph_stable-4.10
----

> * For Operator updates, you have to do the following:

Reviewer 2: ```suggestion
* For Operator updates, must perform the following tasks:
```

+

> . Mirror the desired index image, bundle images, and all Operator images referenced in the bundle images.

Reviewer 2: ```suggestion
. Mirror the desired index image, the bundle images, and all the Operator images that are referenced in the bundle images.
```

== Preparing for the updates

If both the hub and the spoke clusters are running {product-title} 4.9, you need to update ZTP from version 4.9 to 4.10. If {product-title} 4.10 is used, you can set up the environment.

[id="talo-platform-prepare-for-update-env-setup_{context}"]
== Setting up the environment

If you deployed ZTP in a disconnected environment as instructed in the "Preparing the disconnected environment" section in "Deploying distributed units at scale in a disconnected environment", you need to do the following steps before TALO can successfully perform update operations:

TALO can perform both platform and Operator updates:

* For platform updates, you have to do the following:
+
. Mirror the desired {product-title} image repository. For more information about how to mirror an image repository, see the "Mirroring the OpenShift Container Platform image repository" section.
+
[IMPORTANT]
====
The `imageContentSources` section, which can be retrieved with the `oc adm release mirror` command, needs to be saved for platform updates.
For example, this is the output of the `imageContentSources` which contains the information of your mirrors:
+
[source,terminal]
----

> imageContentSources:

Reviewer 5: The out put must include two parts. Usually something like below (change mirror names please). Also please fix the indentation in the example.

```
imageContentSources:
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-release
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
```

       - mirrors:
         - registry.example.com:5000/openshift4
         source: quay.io/openshift-release-dev/ocp-v4.0-art
----
====

. Prepare the update graph. You have two options to prepare the update graph:

> * Use the OpenShift Update Service.

Reviewer 5: "Use the OpenShift Update Service" and "Make a local copy of the upstream graph.. "  are two options for the "Prepare the update graph" section above. Right now they look like they are outside of "Prepare the update graph".

[id="talo-platform-prepare-end-to-end_{context}"]

> = End-to-end procedures for updating clusters in a disconnected environment

Reviewer 5: This section and its contents should be moved out one level instead of under "Updating managed policies with the Topology Aware Lifecycle Operator".

If you have deployed spoke clusters with distributed unit (DU) profiles using the GitOps ZTP with Topology Aware Lifecycle Operator pipeline described in "Deploying distributed units at scale in a disconnected environment", this procedure describes how to upgrade your spoke clusters and Operators.

[id="talo-platform-prepare-for-update_{context}"]
== Preparing for the updates

If both the hub and the spoke clusters are running {product-title} 4.9, you must update ZTP from version 4.9 to 4.10. If {product-title} 4.10 is used, you can set up the environment.

[id="talo-platform-prepare-for-update-env-setup_{context}"]
== Setting up the environment

The TALO can perform both platform and Operator updates.

You must mirror both the platform image and Operator images that you want to update to in your mirror registry before you can use the TALO to update your disconnected clusters. Complete the following steps to mirror the images:

* For platform updates, you must perform the following steps:
+

> . Mirror the desired {product-title} image repository. For more information about how to mirror an image repository, see the "Mirroring the OpenShift Container Platform image repository" section.

Reviewer 5: "Mirroring the OpenShift Container Platform image repository" should be changed to a link to existing doc. Also,
the linked instruction is about how to mirror the ocp image repository, so our description should also be updated to reflect that. 
Please see Angie's doc for more details.

[Mirroring the Openshift Container Platform image repository](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html#installation-mirror-repository_installing-mirroring-installation-images)

Author: > "Mirroring the OpenShift Container Platform image repository" should be changed to a link to existing doc

@yliu127 Unfortunately, we can't link there due to technical limitations, but I'll update/add a link in the `Additional resources` after the section.

Reviewer 5: [Mirroring the Openshift Container Platform image repository](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html#installation-mirror-repository_installing-mirroring-installation-images) is from an existing ocp doc. And I saw links in procedures in this doc: https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html#installation-mirror-repository_installing-mirroring-installation-images
Can we do the same?

> . Save the output of the `imageContentSources` section in a file. Retrieve the contents of the `imageContentSources` section, by running the following command:

Reviewer 5: This description is a bit misleading, because the "oc adm release mirror ...." command was already executed by following step #1 to mirror the ocp image. The image content sources to be recorded is part of the output of that command.

+
[source,terminal]
----
$ oc adm release mirror
----
+
.Example output

[source,terminal]
----
imageContentSources:
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-release
 - mirrors:
   - mirror-ocp-registry.ibmcloud.io.cpak:5000/openshift-release-dev/openshift4
   source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
----
+

> [NOTE]

Reviewer 5: This should not be under "Save the output of the `imageContentSources` section in a file ... ", they are in parallel. From this line to line 104, the steps are about how to get the image signature of an ocp image.

----

.. Specify the architecture of the server by running the following command:
+
[source,terminal]
----
$ ARCHITECTURE=<server_architecture>
----

.. Get the release image digest from Quay by running the following command
+
[source,terminal]
----
$ DIGEST="$(oc adm release info quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE_NUMBER}-${ARCHITECTURE} | sed -n 's/Pull From: .*@//p')"
----

.. Set the digest algorithm by running the following command:
+
[source,terminal]
----
$ DIGEST_ALGO="${DIGEST%%:*}"
----

.. Set the digest signature by running the following command:
+
[source,terminal]
----
$ DIGEST_ENCODED="${DIGEST#*:}"
----

.. Get the image signature from the link:https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/[mirror.openshift.com] website by running the following command:
+
[source,terminal]
----
$ SIGNATURE_BASE64=$(curl -s "https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${DIGEST_ALGO}=${DIGEST_ENCODED}/signature-1" | base64 -w0 && echo)
----

.. Save the image signature to the `checksum-<OCP_RELEASE_NUMBER>.yaml` file by running the following commands:
+
[source,terminal]
----

> $ cat >checksum-${OCP_RELEASE_NUMBER}.yaml <<EOF

Reviewer 5: Following is meant to be one command:

cat >checksum-${OCP_RELEASE_NUMBER}.yaml <<EOF
${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATURE_BASE64}
EOF

----
+
[source,terminal]
----

> $ {DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATURE_BASE64}

Reviewer 5: See my comment above.

----

. Prepare the update graph. You have two options to prepare the update graph:

.. Use the OpenShift Update Service.
+

> For more information about how to set up the graph on the hub cluster, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#build-the-graph-data-init-container[Build the graph data init container].

Reviewer 5: We need to use both of the following references to set up upgrade graph using OpenShift Update Service: 

https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#deploy-the-operator-for-cincinnati

and 

https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.4/html/clusters/managing-your-clusters#build-the-graph-data-init-container

.. Make a local copy of the upstream graph. Host the update graph on an `http` or `https` server in the disconnected environment that has access to the spoke cluster. To download the update graph, use the following command:
+
[source,terminal]
----
$ curl -s https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.10 -o ~/upgrade-graph_stable-4.10
----

* For Operator updates, you must perform the following tasks:
+

> . Mirror the Operator catalogs. Ensure that the desired operator images are mirrored by following the procedure in the "Mirroring Operator catalogs for use with disconnected clusters" section.

Reviewer 5: "Mirroring Operator catalogs for use with disconnected clusters" should be changed to links to exiting docs. Please see Angie's doc for more details.

+

> For more information about how to mirror images, see the "Preparing the disconnected environment" section.

Reviewer 5: This should be removed because it's already covered under "Additional resources"
